### ISSUE TITLE
How to display image in the window

### ISSUE THREAD

**Issue by andreykaere** [psychon/x11rb#795]
Hi! I am trying to get this code working
```rust
use std::error::Error;

use image::io::Reader;
use image::GenericImageView;
use x11rb::connection::Connection;
use x11rb::protocol::xproto::*;
use x11rb::protocol::Event;

pub fn display_image(
    image_path: &str,
    x: u16,
    y: u16,
    size: u16,
) -> Result<(), Box<dyn Error>> {
    let image = Reader::open(image_path)?.decode()?;
    let (width, height) = image.dimensions();

    let (conn, screen_num) = x11rb::connect(None)?;
    let screen = &conn.setup().roots[screen_num];

    let win = conn.generate_id()?;
    let window_aux = CreateWindowAux::new()
        .border_pixel(screen.black_pixel)
        .event_mask(EventMask::EXPOSURE);

    create_window(
        &conn,
        x11rb::COPY_FROM_PARENT as u8,
        win,
        screen.root,
        x.try_into()?,
        y.try_into()?,
        width.try_into()?,
        height.try_into()?,
        0,
        WindowClass::COPY_FROM_PARENT,
        screen.root_visual,
        &window_aux,
    )?;

    let gc_aux = CreateGCAux::new();
    let gc = conn.generate_id()?;
    create_gc(&conn, gc, win, &gc_aux)?;

    let pixmap = conn.generate_id()?;
    conn.create_pixmap(
        screen.root_depth,
        pixmap,
        win,
        width as u16,
        height as u16,
    )?;

    let data = image.into_rgba8().into_raw();
    conn.put_image(
        ImageFormat::Z_PIXMAP,
        pixmap,
        gc,
        width as u16,
        height as u16,
        0,
        0,
        0,
        screen.root_depth,
        &data,
    )?;

    conn.copy_area(
        pixmap,
        win,
        gc,
        0,
        0,
        0,
        0,
        width.try_into()?,
        height.try_into()?,
    )?;

    conn.map_window(win)?;
    conn.flush()?;

    loop {
        let event = conn.wait_for_event()?;

        if let Event::Expose(_) = event {
            conn.flush()?;
        }
    }
}
```

However, it just displays window and that's it. I would really appreciate, if you help me to fix it. 

By the way, could you kindly explain to me, why do we need infinite loop of event for window to be displayed? Why if we don't do it, it won't be displayed (or will be killed right after displayed)?

Thank you in advance!

**Comment by psychon** [comment]
>  I would really appreciate, if you help me to fix it.

You are close to it. The reason it does not work is that you are trying to draw to the window before it is visible. X11 does not keep window content around when it is not visible. There is just the stuff that is shown on screen. If something becomes newly visible, you get an Expose event and have to handle it by redrawing the affected parts. It "works" for me like this:
```diff
diff --git a/src/main.rs b/src/main.rs
index c18d1fc..d00a8a7 100644
--- a/src/main.rs
+++ b/src/main.rs
@@ -70,18 +70,6 @@ pub fn display_image(
         &data,
     )?;
 
-    conn.copy_area(
-        pixmap,
-        win,
-        gc,
-        0,
-        0,
-        0,
-        0,
-        width.try_into()?,
-        height.try_into()?,
-    )?;
-
     conn.map_window(win)?;
     conn.flush()?;
 
@@ -89,6 +77,17 @@ pub fn display_image(
         let event = conn.wait_for_event()?;
 
         if let Event::Expose(_) = event {
+            conn.copy_area(
+                pixmap,
+                win,
+                gc,
+                0,
+                0,
+                0,
+                0,
+                width.try_into()?,
+                height.try_into()?,
+            )?;
             conn.flush()?;
         }
     }
```
However, now the colors are off. Pixel formats in X11 are complicated and the image crate does not seem to have anything helpful for this conversion. The short version to get this to work is:
```diff
diff --git a/src/main.rs b/src/main.rs
index d00a8a7..6d52e79 100644
--- a/src/main.rs
+++ b/src/main.rs
@@ -56,7 +56,12 @@ pub fn display_image(
         height as u16,
     )?;
 
-    let data = image.into_rgba8().into_raw();
+    let mut data = image.into_rgba8().into_raw();
+    data.chunks_exact_mut(4).for_each(|chunk| {
+        let (c0, c2) = (chunk[0], chunk[2]);
+        chunk[2] = c0;
+        chunk[0] = c2;
+    });
     conn.put_image(
         ImageFormat::Z_PIXMAP,
         pixmap,
```
The long version goes something like "X11 has the concept of visuals that describe how pixels values are interpreted" and "please look at https://github.com/psychon/x11rb/blob/master/x11rb/examples/display_ppm.rs for an example".

> By the way, could you kindly explain to me, why do we need infinite loop of event for window to be displayed? Why if we don't do it, it won't be displayed (or will be killed right after displayed)?

Why the window will be destroyed: When you exit your function, the X11 connection is closed (`conn` is dropped). When the X11 connection is closed, the X11 server cleans up and destroys everything you created, including the window.

And why you need the loop is hopefully explained by my patch above: Something needs to handle events like requests to redraw parts of the window or input events.

Basically: Because nothing is left that handles the input / deals with it.

**Comment by andreykaere** [comment]
Thank you for your comprehensive answers. However I still have some questions:
1) Why is image quality so low? For example, if I open the same file in inkscape, it looks a lot more nicer. I suspect, it's because of the way pixmap works. But at the end of the day inkscape also makes some calls to xorg and displays its content via the same technique. Where's the trick?:)
2) Why is that better " I changed the code so that the pixmap is set as the background of your window.", than not having a background at all? 
3) How to make background transparent? (I want to display png with transparent background)

**Comment by psychon** [comment]
> Why is image quality is not that hight? 

Uhm. I don't know. While debugging the swapped colors, I opened a JPEG with `feh` and your program and switched between the two. I couldn't notice any difference between the two images after fixing the colors. So... random guess would be "something with the `image` crate"? Are you sure this happens for all images? If so, perhaps you could post your test image and a screenshot?

> Why is that better " I changed the code so that the pixmap is set as the background of your window.", than not having a background at all?

Well, you (basically) asked for a way to show some image without having to handle events. By setting the pixmap as the background, the X11 server draws it and the program does not have to handle expose events. But I agree that this is in general a bad idea, especially as soon as one stops displaying a simple, static image.

> How to make background transparent? (I want to display png with transparent background)

Short version: It's complicated. If possible, don't try to go there.

Long version:

X11 by itself does not have any concept of transparency. This only became possible with the RENDER and COMPOSITE extensions and requires an extra program (the composite manager) to be running, although these days most desktop environments have one built-in.

The [cairo example](https://github.com/psychon/x11rb/tree/master/cairo-example) shows how to do it.
Basically:
- The RENDER extension introduces something called "picture format". This is basically the same as a visual (describes some color format), but also supports transparency
- There is a mapping between picture formats and visuals and the specification requires an ARGB32 picture format to always be present
- The ARGB32 picture format as to be found and the corresponding visual is used for creating a window (see `choose_visual()` in the cairo example)
- The RENDER and COMPOSITE extensions are extensions, so could (in theory) not be supported by the X11 server. This is handled via e.g. `conn.extension_information(render::X11_EXTENSION_NAME)?.is_some()`
- Composite managers are optional, so we have to check whether one is running (and in theory even have to handle the cases where a composite manager is starting / stopping while our program is already running, but I guess in practice this extra trouble is not worth it) (see function `composite_manager_running`)
- Once all of this is done, it *should* be possible to just upload pixel data in ARGB32 format (basically: What the image crate gives you with `into_argb8()`) with `PutImage` just as your original program does

**Comment by psychon** [comment]
> Why blue and red colors come swapped from x11 and to put it into x11 we have to swap them as well?

X11 is quite flexible in how it can work with colors. The details about how colors are represented are described in a visual. These days, most X11 server have a similar visual as the default visual. "Back in the day", things where complicated.

Anyway, let's look at my current X11 server. `xdpyinfo` is a tool that dumps the information that you can get from `conn.setup()` in a human readable format.

First, what is the number of the default visual? In x11rb, this is something like `setup().roots[0].root_visual`
```
$ xdpyinfo | grep 'default visual id'
  default visual id:  0x21
```
This 0x21 is just an arbitrary number. What does it mean? In x11rb, one has to dig through `setup().roots[0].allowed_depths` and its contained values to get this information (because X11 represents things this way..).
```
$ xdpyinfo | grep -A 5 0x21
  default visual id:  0x21
  visual:
    visual id:    0x21
    class:    TrueColor
    depth:    24 planes
    available colormap entries:    256 per subfield
    red, green, blue masks:    0xff0000, 0xff00, 0xff
    significant bits in color specification:    8 bits
```
So, this visual has a depth of 24 pixels. TrueColor basically means that colors are represented the way that you expect it to. The most significant byte is red, then comes green, then blue (this info comes from the `masks`).

Last piece of information (x11rb: `setup().pixmap_formats`):
```
$ xdpyinfo| grep 'depth 24' 
    depth 24, bits_per_pixel 32, scanline_pad 32
```
So for a visual with depth 24, pixels are arranged into 32 bit quantities.

To summarize:
- Each pixel is a u32
- With color order rgb
- Thanks to little endian, this means that the order of the components in-memory is bgrX (where X is unused)

And for the image crate: `into_rgba8()` produces a series of u8 values, where each group of 4 describes a pixel with the first being the red component, the next one green etc.

To turn rgbX into bgrX, one has to swap the first and third byte, which the code above does.

Now... that's the part about how to figure out this stuff. As to why the X11 server chooses to represent colors this way: I don't know. But e.g. the cairo library also does the same (32 bit quantity in native endianness): https://cairographics.org/manual/cairo-Image-Surfaces.html#cairo-format-t

So, to me the question is rather: Why does the image crate represent a color as a series of bytes instead of a single quantity in native endianness? But I guess in the end that's a matter of taste...

**Comment by andreykaere** [comment]
Idk, to me representing RGB in the ordered rgb is quite natural:) But the question is then, I guess, why it was chosen to use little endian and this order of bytes, such that red and blue are getting swapped ... I mean, I would guess that the answer might be "it's historical thing". But the fact that cairo also uses makes me wondering if there is some more profound reason for it ...

**Comment by psychon** [comment]
> But the fact that cairo also uses makes me wondering if there is some more profound reason for it ...

AFAIK cairo started as a library called Xr. Google found https://www.cairographics.org/xr_ols2003/. From that page:

> Xr provides a vector-based rendering API with output support for the X Window System and local image buffers. PostScript and PDF file output is planned. 

So I would _guess_ that cairo inherited this from X11.

**Comment by andreykaere** [comment]
So, it means that this is more or less just the case for X11? (switching red and blue)

**Comment by psychon** [comment]
I don't know. I am only really around cairo / X11. I know that gdk-pixbuf also has an endian-independent pixel format (unlike cairo / X11). Uhm.... that's about the only other library that comes to my mind right now.
